login/service.go

```diff
@@ -206,8 +206,65 @@ func (keycloak *KeycloakOAuthProvider) CreateKeycloakUser(accessToken string, ct
 		log.Error(ctx, map[string]interface{}{
 			"keycloakIdentityID": keycloakIdentityID,
 			"err":                err,
-		}, "unable to  query for an identity")
-		return nil, nil, errors.New("Error during querying for an identity " + err.Error())
+		}, "unable to  query for an identity by ID")
+		return nil, nil, errors.New("Error during querying for an identity by ID " + err.Error())
+	}
+
+	if len(identities) == 0 {
+		// No Idenity with the keycloak user ID is found, try to search by the username
+		identities, err = keycloak.Identities.Query(account.IdentityFilterByUsername(claims.Username), account.IdentityWithUser())
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"keycloakIdentityUsername": claims.Username,
+				"err": err,
+			}, "unable to  query for an identity by username")
+			return nil, nil, errors.New("Error during querying for an identity by username " + err.Error())
+		}
+		if len(identities) != 0 {
+			idn := identities[0]
+			if idn.ProviderType == account.KeycloakIDP {
+				// This is not what actaully should happen.
+				// This is a workaround for Keyclaok and DB unsynchronization.
+				// The old identity will be removed. The new one with proper ID will be created.
+				// All links to the old identities (in Work Items for example) will still point to the deleted identity.
+				log.Warn(ctx, map[string]interface{}{
+					"keycloakIdentityID":       keycloakIdentityID,
+					"coreIdentityID":           idn.ID,
+					"keycloakIdentityUsername": claims.Username,
+				}, "the identity ID fetched from Keycloak and the identity ID from the core DB for the same username don't match. The identity will be re-created.")
+
+				err = application.Transactional(keycloak.db, func(appl application.Application) error {
+					user = &idn.User
+					identity = &account.Identity{
```

Comment:
I'm not a fan of this either. This is just a simplest workaround. Switching to usernames will require much bigger change because it will require us to load an identity from DB everytime we need to find a user/idenitity out of a token.
We could think about a better option (if we really need it) later.